driver/sync_progress_tracker.go

```diff
@@ -0,0 +1,250 @@
+			return
+		}
+
+		if new(big.Int).SetUint64(headHeight).Cmp(s.lastSyncedVerifiedBlockHeight) >= 0 {
+			s.lastProgressedTime = time.Now()
+			log.Info("L2 execution engine has finished the P2P sync work, all verfiied blocks synced, "+
+				"will switch to insert pending blocks ony be one",
+				"lastSyncedVerifiedBlockID", s.lastSyncedVerifiedBlockID,
+				"lastSyncedVerifiedBlockHeight", s.lastSyncedVerifiedBlockHeight,
+				"lastSyncedVerifiedBlockHash", s.lastSyncedVerifiedBlockHash,
+			)
+			return
+		}
+
+		log.Warn("L2 execution engine has not started P2P syncing yet")
+	}
+
+	defer func() { s.lastSyncProgress = progress }()
+
+	// Check whether the L2 execution engine has synced any new block through P2P since last event loop.
+	if syncProgressed(s.lastSyncProgress, progress) {
+		s.outOfSync = false
+		s.lastProgressedTime = time.Now()
+		return
+	}
+
+	// Has not synced any new block since last loop, check whether reaching the timeout.
+	if time.Since(s.lastProgressedTime) > s.timeout {
+		// Mark the L2 execution engine out of sync.
+		s.outOfSync = true
+
+		log.Warn(
+			"L2 execution engine out of sync",
+			"lastProgressedTime", s.lastProgressedTime,
+			"timeout", s.timeout,
+		)
+	}
+}
+
+// UpdateMeta updates the inner beacon sync meta data.
+func (s *BeaconSyncProgressTracker) UpdateMeta(id, height *big.Int, blockHash common.Hash) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	log.Debug("Update sync progress tracker meta", "id", id, "height", height, "hash", blockHash)
+
+	if !s.triggered {
+		s.lastProgressedTime = time.Now()
+	}
+
+	s.triggered = true
+	s.lastSyncedVerifiedBlockID = id
+	s.lastSyncedVerifiedBlockHeight = height
+	s.lastSyncedVerifiedBlockHash = blockHash
+}
+
+// ClearMeta cleans the inner beacon sync meta data.
+func (s *BeaconSyncProgressTracker) ClearMeta() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	log.Debug("Clear sync progress tracker meta")
+
+	s.triggered = false
+	s.lastSyncedVerifiedBlockID = nil
+	s.lastSyncedVerifiedBlockHash = common.Hash{}
+	s.outOfSync = false
+}
+
+// HeadChanged checks if a new beacon sync request will be needed.
+func (s *BeaconSyncProgressTracker) HeadChanged(newID *big.Int) bool {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if !s.triggered {
+		return true
+	}
+
+	return s.lastSyncedVerifiedBlockID != nil && s.lastSyncedVerifiedBlockID != newID
+}
+
+// OutOfSync tells whether the L2 execution engine is marked as out of sync.
+func (s *BeaconSyncProgressTracker) OutOfSync() bool {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.outOfSync
+}
+
+// Triggered returns tracker.triggered.
+func (s *BeaconSyncProgressTracker) Triggered() bool {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.triggered
+}
+
+// LastSyncedVerifiedBlockID returns tracker.lastSyncedVerifiedBlockID.
+func (s *BeaconSyncProgressTracker) LastSyncedVerifiedBlockID() *big.Int {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.lastSyncedVerifiedBlockID
```

Comment:
maybe return a copy instead of pointer to avoid data-race

Reply:
I think there wont be a data-race because of the `Rlock()` in line 191 ?

Reply:
changed: https://github.com/taikoxyz/taiko-client/pull/61/commits/12fb3cc0fc13cef0ccdbfe23d178f47fde2e73a4